File: architecture-skills/design-patterns/examples/bridge/bridge.js
// Bridge pattern in JavaScript.
//
// Shapes hold a renderer object and delegate to it. With duck typing
// the "implementation interface" is just a convention: any object with
// circle() and rectangle() methods works as a renderer.

class VectorRenderer {
  circle(radius) {
    return `<circle r=${radius.toFixed(1)}/>`;
  }
  rectangle(width, height) {
    return `<rect w=${width.toFixed(1)} h=${height.toFixed(1)}/>`;
  }
}

class RasterRenderer {
  circle(radius) {
    return `pixels for a circle of radius ${radius.toFixed(1)}`;
  }
  rectangle(width, height) {
    return `pixels for a ${width.toFixed(1)}x${height.toFixed(1)} rectangle`;
  }
}

class Shape {
  constructor(renderer) {
    this.renderer = renderer;
  }
}

class Circle extends Shape {
  constructor(renderer, radius) {
    super(renderer);
    this.radius = radius;
  }
  draw() {
    return this.renderer.circle(this.radius);
  }
}

class Rectangle extends Shape {
  constructor(renderer, width, height) {
    super(renderer);
    this.width = width;
    this.height = height;
  }
  draw() {
    return this.renderer.rectangle(this.width, this.height);
  }
}

for (const renderer of [new VectorRenderer(), new RasterRenderer()]) {
  for (const shape of [new Circle(renderer, 5.0), new Rectangle(renderer, 3.0, 4.0)]) {
    console.log(shape.draw());
  }
}

File: architecture-skills/design-patterns/examples/bridge/bridge.py
"""Bridge pattern in Python.

Splits shapes (the abstraction) from renderers (the implementation) so
the two hierarchies vary independently: adding a shape never touches
the renderers, and adding a renderer never touches the shapes. With
duck typing the Renderer ABC is optional documentation.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Renderer(ABC):
    """Implementation interface."""

    @abstractmethod
    def circle(self, radius: float) -> str: ...

    @abstractmethod
    def rectangle(self, width: float, height: float) -> str: ...


class VectorRenderer(Renderer):
    def circle(self, radius: float) -> str:
        return f"<circle r={radius:.1f}/>"

    def rectangle(self, width: float, height: float) -> str:
        return f"<rect w={width:.1f} h={height:.1f}/>"


class RasterRenderer(Renderer):
    def circle(self, radius: float) -> str:
        return f"pixels for a circle of radius {radius:.1f}"

    def rectangle(self, width: float, height: float) -> str:
        return f"pixels for a {width:.1f}x{height:.1f} rectangle"


class Shape(ABC):
    """Abstraction: holds a renderer and delegates the drawing to it."""

    def __init__(self, renderer: Renderer) -> None:
        self.renderer = renderer

    @abstractmethod
    def draw(self) -> str: ...


class Circle(Shape):
    def __init__(self, renderer: Renderer, radius: float) -> None:
        super().__init__(renderer)
        self.radius = radius

    def draw(self) -> str:
        return self.renderer.circle(self.radius)


class Rectangle(Shape):
    def __init__(self, renderer: Renderer, width: float, height: float) -> None:
        super().__init__(renderer)
        self.width = width
        self.height = height

    def draw(self) -> str:
        return self.renderer.rectangle(self.width, self.height)


if __name__ == "__main__":
    for renderer in (VectorRenderer(), RasterRenderer()):
        for shape in (Circle(renderer, 5.0), Rectangle(renderer, 3.0, 4.0)):
            print(shape.draw())
